src/mrpro/algorithms/_prewhiten_kspace.py

```diff
@@ -0,0 +1,61 @@
+"""Prewhiten k-space data."""
+
+# Copyright 2023 Physikalisch-Technische Bundesanstalt
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#       http://www.apache.org/licenses/LICENSE-2.0
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from __future__ import annotations
+
+import numpy as np
+import torch
+from einops import rearrange
+
+from mrpro.data import KData
+from mrpro.data import KNoise
+
+
+def prewhiten_kspace(kdata: KData, knoise: KNoise, scale_factor: float = 1.0) -> KData:
+    """Calculate noise prewhitening matrix and decorrelate coils.
+
+    This function is inspired by https://github.com/ismrmrd/ismrmrd-python-tools.
+
+    More information can be found in
+    http://onlinelibrary.wiley.com/doi/10.1002/jmri.24687/full
+    https://doi.org/10.1002/nbm.1429
+
+    Parameters
+    ----------
+    kdata
+        K-space data.
+    knoise
+        Noise measurements.
+    scale_factor, optional
+        Square root is applied on the noise covariance matrix. Used to adjust for effective noise bandwith
+        and difference in sampling rate between noise calibration and actual measurement:
+        scale_factor = (T_acq_dwell/T_noise_dwell)*NoiseReceiverBandwidthRatio, by default 1.0
+
+    Returns
+    -------
+        Prewhitened k-space data
+    """
+    # Reshape noise to (coil, everything else)
+    noise = rearrange(knoise.data, 'other coils k2 k1 k0->coils (other k2 k1 k0)')
+
+    # Calculate noise covariance matrix which should ideally be a unity matrix, i.e. no noise correlation between coils
+    noise_cov = (1.0 / (noise.shape[1])) * torch.einsum('ax,bx->ab', noise, noise.conj())
+
+    # Calculate prewhitening matrix and scale
+    prew = torch.linalg.inv(torch.linalg.cholesky(noise_cov)) * np.sqrt(scale_factor)
+
+    # Apply prewhitening matrix
+    prew_data = torch.einsum('yx,axbcd->aybcd', prew, kdata.data)
```

Comment:
https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/

Reply:
> https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/ ;)

Great post!

`cholesky_solve` does not work but `torch.linalg.solve_triangular` would work. Nevertheless, then I don't see a way of doing the einsum application to the data but would need to rearrange to (other, coil, (k2,k1,k0)), solve and rearrange back. Seems a bit overkill for a matrix size which is never more than 40 x 40.

Reply:
ups. I completely misunderstood to goal of these lines. Sorry.